src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx

```diff
@@ -290,70 +295,105 @@ export const useSendDetails = (): UseSendDetailsReturnType => {
     }
   }
 
-  const inputHandler = useCallback(
+  /**
+   * handleInputChange
+   *
+   * Determines the form's state from input by onChange event.
+   * Valid inputs:
+   * - Non-empty numeric values including zero
+   * Error states:
+   * - Insufficient funds - give > have
+   * - Empty amount - input = ''
+   * - Not enough native token - gas > have
+   */
+  const handleInputChange = useCallback(
     async (inputValue: string) => {
-      setLoading(true)
       setValue(SendFormFields.SendMax, false)
+
       const key =
         fieldName !== SendFormFields.FiatAmount
           ? SendFormFields.FiatAmount
           : SendFormFields.CryptoAmount
-      if (inputValue === '') {
-        // Don't show an error message when the input is empty
-        setValue(SendFormFields.AmountFieldError, '')
-        setLoading(false)
-        // Set value of the other input to an empty string as well
-        setValue(key, '')
-        return
-      }
-      const amount =
-        fieldName === SendFormFields.FiatAmount
-          ? bnOrZero(bn(inputValue).div(price)).toString()
-          : bnOrZero(bn(inputValue).times(price)).toString()
 
-      setValue(key, amount)
+      try {
+        if (inputValue === '') {
+          // TODO: use isDirty instead
+          // Cancel any pending requests
+          debouncedEstimateFormFees.cancel()
+          // Don't show an error message when the input is empty
+          setValue(SendFormFields.AmountFieldError, '')
+          // Set value of the other input to an empty string as well
+          setValue(key, '') // TODO: this shouldnt be a thing, using a single amount field
+          return
+        }
 
-      let estimatedFees
+        const amount =
+          fieldName === SendFormFields.FiatAmount
+            ? bnOrZero(bn(inputValue).div(price)).toString()
+            : bnOrZero(bn(inputValue).times(price)).toString()
+
+        setValue(key, amount)
+
+        // TODO: turn to actual validators
+        let estimatedFees
+        let hasEnoughNativeTokenForGas = false
+        let hasValidBalance = false
+
+        try {
+          setLoading(true)
+          estimatedFees = await debouncedEstimateFormFees()
```

Comment:
Agreed. There's no reason to go out and calculate fees if it is obvious that the amount entered is greater than the user's available balance.

I moved the logic for hasValidBalance above the line to estimate fees and it is working. It only estimates fees now if the amount entered is less than the balance.

I also plugged a hole that if the user is sending the native fee asset and they enter an amount less than their balance, but when including fees is over their balance, it will show insufficient funds. Before it was allowing them to continue and then erroring on confirm.

Reply:
Tested it out and that feels much smoother now with that early check without a network call.

WRT gas estimation issues, this is a long-running issue that was tackled multiple times as bounties, looks like a regression that https://github.com/shapeshift/web/issues/1966 should finally solve across the app.
Tested this one out and it seems to solve it as well, however, I noticed some weirdnesses with error handling that stays in errored/valid state when it shouldn't:

https://user-images.githubusercontent.com/17035424/178613587-630941d2-3f57-47ea-9e86-b91921f173a4.mov